Palette

chkngit opened the pull request:
**Description**
Palette Prediction support

Closes #744

**Author**
@chkngit 

**Type of change**
Enhancement

**Tests and performance**
BDR gain (on SC clips) : ~-3.4% (using 811 PSNR/SSIM)

Speed Loss:
on pure SC clips :  -9% ( on Xeon 8180 )  -4%(on AWS-48 cores) 
on mixed SC/non-SC clips : no impact (on AWS-48 cores)
Memory :+12% ( on Xeon 6148 )

hassount commented on Source/Lib/Common/Codec/EbCodingLoop.c:
```diff
@@ -2382,6 +2402,17 @@ EB_EXTERN void av1_encode_pass(
                         context_ptr->intra_coded_area_sb[tbAddr] += blk_geom->bwidth* blk_geom->bheight;
+
+
+
```
extra spaces

hassount commented on Source/Lib/Common/Codec/EbEntropyCoding.c:
```diff
@@ -1323,9 +1323,17 @@ static void EncodeSkipCoeffAv1(
+#if PAL_SUP
+      palette_size == 0 &&
+#else
+       //  mbmi->palette_mode_info.palette_size[0] == 0 &&
```
is this meant to be?

chkngit replied:
yes, it ensures that filter intra is not allowed when palette is on

hassount commented on Source/Lib/Common/Codec/EbModeDecision.c:
```diff
@@ -5372,6 +5397,139 @@ void  inject_filter_intra_candidates(
+
+    for(int cd=0;cd<14;cd++)
+        palette_cand_array[cd].color_idx_map = malloc(64 * 64);
+
```
do we need this ?

chkngit replied:
some scratch buffers to be  used for internal palette work, I will use the macro for the size though

hassount commented on Source/Lib/Common/Codec/EbModeDecision.c:
```diff
@@ -5372,6 +5397,139 @@ void  inject_filter_intra_candidates(
+        memcpy(&candidateArray[canTotalCnt].palette_info.pmi, &palette_cand_array[cand_i].pmi, sizeof(PALETTE_MODE_INFO));
+#else
+        memcpy(&candidateArray[canTotalCnt].palette_info, &palette_cand_array[cand_i], sizeof(PALETTE_INFO));
+#endif
```
Please remove the #if 1

chkngit replied:
ok

hassount commented on Source/Lib/Common/Codec/EbModeDecision.c:
```diff
@@ -5372,6 +5397,139 @@ void  inject_filter_intra_candidates(
+    }
+
+
+    for (int cd = 0; cd < 14; cd++)
```
too many extra spaces

hassount commented on Source/Lib/Common/Codec/EbProductCodingLoop.c:
```diff
@@ -1692,6 +1703,21 @@ void set_md_stage_counts(
 #endif
+#if PAL_CLASS
+    if (picture_control_set_ptr->parent_pcs_ptr->palette_mode == 1)
+        context_ptr->md_stage_3_count[CAND_CLASS_7] =// context_ptr->bypass_stage2[CAND_CLASS_7] ? context_ptr->md_stage_2_count[CAND_CLASS_7] :
```
do we need the commented part?

chkngit replied:
no need

hassount commented on Source/Lib/Common/Codec/EbProductCodingLoop.c:
```diff
@@ -1804,6 +1830,15 @@ void set_md_stage_counts(
+   // stage2_cand_count[CAND_CLASS_i] = bypass_stage2 ? stage3_cand_count[CAND_CLASS_i] : stage2_cand_count[CAND_CLASS_i];
+   // stage1_cand_count[CAND_CLASS_i] = bypass_stage1 ? stage2_cand_count[CAND_CLASS_i] : stage1_cand_count[CAND_CLASS_i];
+
+
```
comments to be removed

chkngit replied:
added those comments be used later as logic and as reminder for a TODO work

hassount commented on Source/Lib/Common/Codec/EbProductCodingLoop.c:
```diff
@@ -1804,6 +1830,15 @@ void set_md_stage_counts(
+   // stage1_cand_count[CAND_CLASS_i] = bypass_stage1 ? stage2_cand_count[CAND_CLASS_i] : stage1_cand_count[CAND_CLASS_i];
+
+
+#if PAL_CLASS  //THIS SHOULD BE rEMOVED AFTER REBAS~~~
```
is this comment still valid ?

chkngit replied:
left there to be changed if md_staging PR will go first, so I ll leave it there and removed after rebasing on top of md staging

hassount commented on Source/Lib/Common/Codec/k_means_template.h:
```diff
@@ -0,0 +1,125 @@
+//#include "av1/encoder/random.h"
+
+
+
```
too many extra spaces + commented code to be removed

hassount commented on Source/Lib/Common/Codec/EbPictureDecisionProcess.c:
```diff
@@ -998,10 +998,29 @@ EbErrorType signal_derivation_multi_processes_oq(
+        picture_control_set_ptr->palette_mode = 0;
+
+
+
```
too many extra spaces

hassount commented on Source/Lib/Common/Codec/EbPictureDecisionProcess.c:
```diff
@@ -998,10 +998,29 @@ EbErrorType signal_derivation_multi_processes_oq(
     }
 
+#if PAL_SUP
+    //Palette Modes:  0:OFF  1:Slow ... 6:Fastest
```
Can you please add more comments on what are the differences between the modes ?
also we need a configuration flag for palette with the settings above:
-1 auto modes (i.e. default configuration) 0 OFF, 1-6 the levels

hassount commented:
@chkngit 
Please update the description field following the rest of the PRs (e.g. #716 ).
Speed results would have to be reported based on a 100% CPU utilization scenario on a cloud instance. 
Can you please give more details on the 3.4% gain ? is it yuv 611 ? 811 ? SSIM PSNR average ? ..

joelsole commented on Source/Lib/Common/Codec/EbDefinitions.h:
```diff
@@ -2311,6 +2323,21 @@ typedef enum EbAsm
+typedef struct {
+    PALETTE_MODE_INFO pmi;
+    uint8_t  *color_idx_map;
+} PALETTE_INFO;
```
Style: this should be CamelCase

chkngit replied:
ok, sounds good

joelsole commented on Source/Lib/Common/Codec/EbDefinitions.h:
```diff
@@ -2311,6 +2323,21 @@ typedef enum EbAsm
+    // Value of base colors for Y, U, and V
+    uint16_t palette_colors[3 * PALETTE_MAX_SIZE];
+    // Number of base colors for Y (0) and UV (1)
+    uint8_t palette_size[2];
```
palette_size is already defined in BlockModeInfo. 
Could these two be harmonized?

chkngit replied:
ok, sounds good

chkngit replied:
actually in the decoder only the size is part of the mbmi, in the decoder they seem to have included the colors to parsing neighboring struct "PArseNbr4x4Ctxt
 , and we need both the colors and the size, hence the usage of paletteModeInfo struct in the encoder!

joelsole commented on Source/Lib/Common/Codec/palette.c:
```diff
@@ -0,0 +1,778 @@
+ static const int palette_color_index_context_lookup[MAX_COLOR_CONTEXT_HASH +
+     1] = { -1, -1, 0, -1, -1,
+            4,  3,  2, 1 };
+
```
There are some definitions already in EbDecParseInterBlock.c

chkngit commented on Source/Lib/Common/Codec/EbDefinitions.h:
```diff
@@ -2311,6 +2323,21 @@ typedef enum EbAsm
+    // Number of base colors for Y (0) and UV (1)
+    uint8_t palette_size[2];
+
+} PALETTE_MODE_INFO;
```
ok, sounds good

andrey-norkin commented on Source/Lib/Common/Codec/EbModeDecision.c:
```diff
@@ -5372,6 +5397,119 @@ void  inject_filter_intra_candidates(
+void  search_palette_luma(
+    PictureControlSet            *picture_control_set_ptr,
+    ModeDecisionContext          *context_ptr,
+    PALETTE_INFO                 *palette_cand,
```
show be CamelCase, as mentioned in the other comment

chkngit replied:
ok, sounds good

andrey-norkin commented on Source/Lib/Common/Codec/EbModeDecision.c:
```diff
@@ -5372,6 +5397,119 @@ void  inject_filter_intra_candidates(
+    PictureControlSet            *picture_control_set_ptr,
+    ModeDecisionContext          *context_ptr,
+    uint32_t                       *candidateTotalCnt) {
+
```
This should be snake_case

andrey-norkin commented on Source/Lib/Common/Codec/EbModeDecision.c:
```diff
@@ -5372,6 +5397,119 @@ void  inject_filter_intra_candidates(
+
+
+
+    uint32_t                  canTotalCnt = *candidateTotalCnt;
```
Should be snake_case

chkngit replied:
ok, sounds good

andrey-norkin commented on Source/Lib/Common/Codec/palette.c:
```diff
@@ -0,0 +1,778 @@
+     ColorCost color_cost;
+ } Av1ColorMapParam;
+
+ static void get_palette_params(FRAME_CONTEXT *frameContext, CodingUnit*cu_ptr, int plane,
```
maybe this should be FrameContext *frame_context

chkngit commented:
> Please, when addressing comments do not force-push a branch, but rather create a new commit and push it instead

this was rebasing + addressing comments; so had to force push; unless there is a way to not do force push if you rebase, then please tell me how to do it

EwoutH commented:
Are all discussions resolve? If so I will run a few benchmarks

hassount commented on Source/Lib/Common/Codec/EbProductCodingLoop.c:
```diff
@@ -6262,6 +6362,10 @@ void move_cu_data_redund(
     CodingUnit *dst_cu){
 
+#if PAL_SUP
+    memcpy(dst_cu->palette_info.color_idx_map, src_cu->palette_info.color_idx_map, MAX_PALETTE_SQUARE);
```
this seems to be still causing a crash when -scm 0

chkngit replied:
should be fixed now

hassount commented on Source/API/EbSvtAv1Enc.h:
```diff
@@ -270,6 +270,11 @@ typedef struct EbSvtAv1EncConfiguration
 
+    /* Palette Mode
+    *
+    * Default is 0. */
```
default seems to be -1 ?

hassount replied:
also, as this is an API change it needs to be added in the user's guide document

chkngit replied:
ok

hassount commented on Source/App/EncApp/EbAppConfig.c:
```diff
@@ -944,6 +948,12 @@ static EbErrorType VerifySettings(EbConfig *config, uint32_t channelNumber)
         return_error = EB_ErrorBadParameter;
     }
 
+    // palette
```
this check needs to move to the library (VerifySettings in EbEncHandle.c) as this is a library feature

chkngit replied:
ok
